File: julia/05-functions/scope.jl
# Variable Scope in Julia

# A global variable
counter = 0

function increment_local()
    counter = 5          # creates a NEW local variable; the global is untouched
    return counter
end

println("Local result: ", increment_local())
println("Global counter: ", counter)

# Use `global` to modify a global variable from inside a function
function increment_global()
    global counter
    counter += 1
    return counter
end

println("After increment_global: ", increment_global())
println("Global counter now: ", counter)

# Local variables are not visible outside the function
function compute()
    temp = 100           # local to compute
    return temp * 2
end

println("compute() = ", compute())

File: pascal/05-functions/scope.pas
program Scope;

var
    counter: Integer;   { global variables, visible everywhere below }
    value: Integer;

{ Pass by value: 'x' is a copy, so the caller is unaffected }
procedure TryToChange(x: Integer);
begin
    x := x * 10;
    WriteLn('Inside procedure, x = ', x);
end;

{ Pass by reference: 'var n' modifies the caller's variable directly }
procedure Increment(var n: Integer);
begin
    n := n + 1;
end;

{ Routines can read and modify global variables }
procedure ShowCounter;
begin
    WriteLn('Global counter = ', counter);
end;

begin
    value := 5;
    TryToChange(value);
    WriteLn('After call, value = ', value);   { still 5 - unchanged }

    counter := 0;
    Increment(counter);
    Increment(counter);
    ShowCounter;                               { now 2 }
end.
